src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseImportBasicService.java

```diff
@@ -0,0 +1,351 @@
+        this.programmingExerciseParticipationService = programmingExerciseParticipationService;
+        this.programmingExerciseTestCaseRepository = programmingExerciseTestCaseRepository;
+        this.staticCodeAnalysisCategoryRepository = staticCodeAnalysisCategoryRepository;
+        this.programmingExerciseRepository = programmingExerciseRepository;
+        this.programmingExerciseService = programmingExerciseService;
+        this.staticCodeAnalysisService = staticCodeAnalysisService;
+        this.auxiliaryRepositoryRepository = auxiliaryRepositoryRepository;
+        this.submissionPolicyRepository = submissionPolicyRepository;
+        this.programmingExerciseTaskRepository = programmingExerciseTaskRepository;
+        this.solutionEntryRepository = solutionEntryRepository;
+    }
+
+    /**
+     * Imports a programming exercise creating a new entity, copying all basic values and saving it in the database.
+     * All basic include everything except for repositories, or build plans on a remote version control server, or
+     * continuous integration server. <br>
+     * There are however, a couple of things that will never get copied:
+     * <ul>
+     *     <li>The ID</li>
+     *     <li>The template and solution participation</li>
+     *     <li>The number of complaints, assessments and more feedback requests</li>
+     *     <li>The tutor/student participations</li>
+     *     <li>The questions asked by students</li>
+     *     <li>The example submissions</li>
+     * </ul>
+     *
+     * @param templateExercise The template exercise which should get imported
+     * @param newExercise      The new exercise already containing values which should not get copied, i.e. overwritten
+     * @return The newly created exercise
+     */
+    @Transactional // ok because we create many objects in a rather complex way and need a rollback in case of exceptions
+    // IMPORTANT: the transactional context only works if you invoke this method from another class
+    public ProgrammingExercise importProgrammingExerciseBasis(final ProgrammingExercise templateExercise, final ProgrammingExercise newExercise) {
+        // Set values we don't want to copy to null
+        setupExerciseForImport(newExercise);
+        newExercise.setBranch(versionControlService.get().getDefaultBranchOfArtemis());
+
+        // Note: same order as when creating an exercise
+        programmingExerciseParticipationService.setupInitialTemplateParticipation(newExercise);
+        programmingExerciseParticipationService.setupInitialSolutionParticipation(newExercise);
+        setupTestRepository(newExercise);
+        programmingExerciseService.initParticipations(newExercise);
+
+        // Hints, tasks, test cases and static code analysis categories
+        Map<Long, Long> newHintIdByOldId = exerciseHintService.copyExerciseHints(templateExercise, newExercise);
+        programmingExerciseRepository.save(newExercise);
```

Comment:
Exercise Hints are currently not allowed for exam exercises, you can simply remove that.

Reply:
This import functionality is used for both exam and course exercises.
It was just a refactoring, so the @Transactional annotation was usable again.